### PR TITLE
fix(adapters): Slack streaming team_id + Teams DM Graph conversation IDs (vercel/chat#330, #403)

### PR DESCRIPTION
**Pull request by patrick-chinchill**
## Summary

Two small upstream bug-fix ports bundled into one PR. Both touch adapter dispatch code and have new regression tests that fail before the fix.

### fix(slack): interactive-payload team_id through streaming context — vercel/chat#330

Slack carries the workspace ID in different shapes depending on the webhook envelope:
- **Message events** (`message`, `app_mention`): top-level `team_id` / `team` (string).
- **Interactive payloads** (`block_actions`, `view_submission`, …): nested `team.id` (object), with `user.team_id` as a final fallback.

The old extraction (`raw.get("team_id") or raw.get("team")`) returned the entire `team` dict for `block_actions`, which then traveled to the Slack adapter as `recipient_team_id` and either crashed Slack streaming API calls or routed them to the wrong workspace.

Moved the extraction into a dedicated `_extract_slack_recipient_team_id` helper in `src/chat_sdk/thread.py` that walks each shape in order and returns `None` when no string ID is found.

### fix(teams): canonical DM conversation ID for Microsoft Graph API — vercel/chat#403

Bot Framework hands out opaque DM conversation IDs (e.g. `a:1xWhatever`) which Graph's `/chats/{chat-id}/messages` endpoint rejects with 404. The canonical Graph chat ID for a 1:1 DM is `19:{userAadId}_{botId}@unq.gbl.spaces`.

- Cache the user's AAD object ID from incoming activities (`from.aadObjectId`) into a new `TeamsDmContext` keyed by base conversation ID.
- Add discriminated union `TeamsGraphContext = TeamsChannelContext | TeamsDmContext`.
- Rename `_get_channel_context` → `_get_graph_context` and add `_chat_id_from_context()` dispatch helper.
- Update `fetch_messages`, `fetch_channel_messages`, and `fetch_channel_info` to dispatch on context type.
- Backwards-compatible: cached entries written before #403 lack a `type` discriminator and are treated as `channel`.

## Tests

- `tests/test_thread_faithful.py`:
  - Parametrized `test_should_pass_stream_options_from_current_message_context` over all four Slack payload shapes (`team_id`, `team` string, `team.id` object, `user.team_id` fallback).
  - New `test_concurrent_block_actions_team_ids_do_not_cross_contaminate` covering hazard #6 (no team_id leak across concurrent requests).
  - New `test_should_forward_structured_stream_chunks_to_adapter_stream_from_an_action_created_thread` (port of upstream's #330 test; restores 0-missing on `thread.test.ts` fidelity).
- `tests/test_teams_coverage.py`: new `TestGraphDmConversationIdResolution` class with 8 cases covering `_chat_id_from_context` (DM / channel / no-context branches), `_cache_user_context` (DM cached / channel skipped / no-aad skipped / DM-like channel adversarial), and end-to-end `fetch_messages` (DM resolves to `19:{aadId}_{botId}@unq.gbl.spaces`, group chat falls back to raw ID).
- Each new test docstring carries a "What to fix if this fails:" line.

Adversarial checks ran per `docs/SELF_REVIEW.md`: pass-interaction (concurrent team_ids) and the "DM-like channel" misclassification both have explicit tests.

## Test plan

- [x] `uv run ruff check src/ tests/ scripts/` — clean
- [x] `uv run ruff format --check src/ tests/ scripts/` — clean
- [x] `uv run python scripts/audit_test_quality.py` — 0 hard failures (39 pre-existing warnings unchanged)
- [x] `TS_ROOT=/tmp/vercel-chat uv run python scripts/verify_test_fidelity.py` — `thread.test.ts` now reports 0 missing
- [x] `uv run pytest tests/ --tb=short -q` — `3681 passed, 2 skipped, 1 failed`
  - Only failure: pre-existing `tests/test_github_webhook.py::TestGitHubAdapterConstructor::test_throws_when_no_auth` (called out as ignorable in the task brief)

## Upstream refs

- vercel/chat#330 — `[chat] fix Slack streaming team ID for interactive payloads` (commit `8a0c7b3`)
- vercel/chat#403 — `fix(adapter-teams): resolve DM conversation IDs for Graph API` (commit `4c24c94`)

https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj

---
_Generated by [Claude Code](https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj)_

### CODE REVIEW

**Review by patrick-chinchill — COMMENTED** [review]
## Code review — vercel/chat#330 + #403 ports

Both ports are tight 1:1 transliterations of upstream `8a0c7b3` / `4c24c94`. Slack helper walk order matches upstream exactly (`team_id` → `team:str` → `team.id:object` → `user.team_id`); Teams discriminated-union dispatch matches the TS version including the legacy-cache fallthrough (`type` absent → treated as channel). The follow-up `_paginate_graph_chat_messages` extraction is behavior-preserving (forward = `gt` + asc + no reverse, backward = `lt` + desc + reverse, `has_more = len >= limit` for both empty and exact-page edges). The concurrency test genuinely interleaves with `asyncio.gather` + `asyncio.sleep(0)` — Hazard #6 is exercised, not just claimed.

### 🟡 Medium

1. **Adversarial gap in `_extract_slack_recipient_team_id` test sweep** (`src/chat_sdk/thread.py` L144‑183, `tests/test_thread_faithful.py` L797‑818). Per `docs/SELF_REVIEW.md` principle #1 (adversarial input sweep), the parametrized test only covers the four happy paths. The helper has explicit guards for shapes that aren't exercised:
   - `raw.team` is a dict but `id` is missing or non-string (e.g. `{"team": {"domain": "x"}}` → must fall through to `user.team_id`)
   - `team_id` / `team` is an empty string `""` (the truthiness guard `and team_id` is the only thing keeping that from being returned — Hazard #1)
   - `raw` is `None` / not a dict at all
   - `user.team_id` is non-string

   Each is a one-line `pytest.mark.parametrize` extension; without them the next refactor of this helper has no signal it broke a fallback.

### 🔵 Nit

2. **`aadObjectId` is interpolated into a Graph URL path with no shape validation** (`src/chat_sdk/adapters/teams/adapter.py` L330). The cached `graph_chat_id` becomes `chat_id` → `https://graph.microsoft.com/v1.0/chats/{chat_id}/messages`. Bot Framework JWT verification authenticates the activity envelope but does not constrain the contents of `from.aadObjectId`. Real Azure AD object IDs are GUIDs, but a malicious / non-conforming activity could inject `/`, `?`, `#`, `:`. The host is hardcoded so this can't be SSRF, and the bearer token is Graph-scoped so worst case is a 404 against a different Graph endpoint — but a one-line `re.fullmatch(r"[0-9a-fA-F-]{36}", aad_object_id)` guard before formatting would make this robust and matches the same defense-in-depth pattern as `_validate_service_url` two functions up.

3. **Backwards-compat claim has no end-to-end regression test for legacy cache shape** (PR body & `_chat_id_from_context` docstring claim "cached entries written before #403 are treated as channel"). `TestGetGraphContext.test_valid_context_from_cache` confirms the legacy `{"team_id": "t1", "channel_id": "c1"}` shape *loads*, but no test asserts it routes through `fetch_messages` to the channel branch (analogous to the new `test_fetch_messages_uses_graph_chat_id_for_dm`). One added test covering legacy cache → `fetch_channel_thread_messages` would lock the claim.

4. **Channel-context cache writes a `"type": "channel"` field that upstream omits** (`src/chat_sdk/adapters/teams/adapter.py` L316). Upstream's TS only sets `type` on the DM context; channel entries are written without the discriminator. Functionally equivalent (both code paths treat `type != "dm"` as channel), but it's a wire-shape divergence that future cross-language deployments could trip on. Either match upstream and drop the key, or add a row to the non-parity table.

### ✅ Looks good

- `_extract_slack_recipient_team_id` walk order is identical to upstream and the parametrized + structured-stream tests faithfully port the upstream `it.each` block.
- Concurrent `block_actions` test is a real concurrency test (`asyncio.gather` + `await asyncio.sleep(0)` interleave point inside `mock_stream`), not the typical sequential-disguised-as-concurrent.
- `_paginate_graph_chat_messages` refactor is semantically identical to the duplicated blocks it replaces; both edge cases (`len == 0` → `False`, `len == limit` → `True`) match upstream and are covered transitively by the existing fetch tests.
- The `is_dm`-prefix guard for DM caching (`not base_channel_id.startswith("19:")`) is tested both positively (DM cached) and adversarially (channel-with-aadObjectId not mis-cached as DM).
- All 14 new tests pass locally.

_Posted by an automated reviewer agent. https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj_

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by patrick-chinchill — COMMENTED** [review]
## Re-review of fixes (round 2)

Walked the four claimed fixes against `f55378a` (chat@4.27.0) upstream commits `8a0c7b3` (#330) and `4c24c94` (#403), then hunted for residual gaps. 299 tests pass in `test_teams_coverage.py` + `test_thread_faithful.py` + `test_coverage_gaps.py`.

### Verified fixes
1. **Adversarial parametrize** (`test_should_pass_stream_options_from_current_message_context`) — 11 cases now: 4 happy paths + 4 fall-through cascades + 3 None-returning rejections (non-dict, empty dict, non-string `team.id`). Order matches the helper's `team_id` → `team` (str) → `team.id` (obj) → `user.team_id` cascade. Solid.
2. **AAD GUID guard** (`_AAD_OBJECT_ID_PATTERN.fullmatch` at adapter.py:348) — `fullmatch` confirmed to reject path-injection (`.../messages`), missing hyphens, non-hex chars, short strings. Test sweeps 6 attack patterns including `../etc/passwd`. Defense-in-depth is real.
3. **Legacy cache E2E** (`test_fetch_messages_uses_legacy_cache_shape_for_channel`) — asserts that pre-#403 entries (no `type` key) load via `_get_graph_context` and `_chat_id_from_context` returns the raw conversation ID rather than a misclassified DM URL. Good backwards-compat coverage.
4. **Wire-shape parity** — channel cache write at adapter.py:322 omits `"type": "channel"`, exactly matching upstream TS index.ts:220. Cross-language wire shape is now identical.

### Hunt for residual upstream parity gaps
- **Race conditions / ContextVar**: `recipient_team_id` flows through per-thread `self._current_message`, no module-global caches. Concurrency test (`test_concurrent_block_actions_team_ids_do_not_cross_contaminate`) verifies two interleaved `block_actions` posts don't cross-pollute. Clean.
- **StreamOptions wire-shape**: walked upstream `chat/src/types.ts` — only `recipientUserId` + `recipientTeamId` are derived from `currentMessage`. No new fields in #330. No gap.
- **Other Graph endpoints consuming `chat_id`**: TS upstream applies `chatIdFromContext` in 3 sites — `fetchMessages`, `fetchChannelMessages`, AND `listThreads`. Python ports the first two correctly. **`list_threads` is not implemented** in the Python Teams adapter at all (pre-existing gap on `main`, not introduced here, but worth tracking — when ported, must thread `_chat_id_from_context` through). `fetch_channel_info` is correctly DM-aware (channel branch only fires when `type != "dm"`).
- **`graph-api.test.ts` `chatIdFromContext` describe block**: all 3 upstream `it()` cases (DM context → `graphChatId`, no context → raw ID, channel context → raw ID) are mirrored as `test_chat_id_from_context_*` in `TestGraphDmConversationIdResolution`. Full fidelity.
- **TS DM-context simplification commit 2** ("DMs never have threadMessageId so the channel guard doesn't need an explicit DM exclusion"): Python's `if graph_context and context_type != "dm" and thread_message_id:` keeps the explicit DM exclusion. Functionally equivalent (more defensive than upstream); no parity bug.

### Nits (non-blocking)
- `TeamsChannelContext(TypedDict, total=False)` makes `team_id` and `channel_id` optional at the type level. Upstream TS only marks `type` optional. Consider `class TeamsChannelContext(TypedDict): team_id: str; channel_id: str; type: NotRequired[str]` for stricter typing. Runtime is unaffected.
- `self._app_id` is interpolated into the Graph chat ID without GUID validation. It's config-sourced (`TEAMS_APP_ID` env), so not user input — but a misconfigured `/`-containing value would silently produce a malformed chat ID. Optional symmetric guard.

### Re-review verdict: PASS

All 4 review-driven fixes land cleanly. Wire-shape, behavior, and test fidelity match upstream. The two nits above are advisory-only.

_Posted by an automated re-reviewer agent. https://claude.ai/code/session_01FyMxQn2BEAzmwKS1GZczKj_

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/chat_sdk/adapters/teams/adapter.py [hunk 1]
```diff
@@ -1176,19 +1205,24 @@ async def fetch_channel_messages(
...
+                        params["$filter"] = f"createdDateTime gt {options.cursor}"
+                    graph_messages = await self._graph_list_chat_messages(chat_id, params)
+                    has_more = len(graph_messages) >= limit
+                else:
+                    params = {"$top": limit, "$orderby": "createdDateTime desc"}
+                    if options.cursor:
+                        params["$filter"] = f"createdDateTime lt {options.cursor}"
+                    graph_messages = await self._graph_list_chat_messages(chat_id, params)
+                    graph_messages.reverse()
+                    has_more = len(graph_messages) >= limit
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/teams/adapter.py]: Addressed in commit cce5c75 — extracted `_paginate_graph_chat_messages(chat_id, limit, direction, cursor) -> (messages, has_more)` and both `fetch_messages` (was L1092-1110) and `fetch_channel_messages` (was L1213-1225) now share a single code path. Behavior is preserved: `gt` + asc + no-reverse on forward, `lt` + desc + reverse on backward, `has_more = len >= limit` for both empty and exact-page edges. Existing `TestGraphDmConversationIdResolution` + `fetch_messages` / `fetch_channel_messages` tests cover both branches.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/chat_sdk/adapters/teams/adapter.py [hunk 2]
```diff
@@ -308,12 +315,44 @@ async def _cache_user_context(self, activity: dict[str, Any]) -> None:
...
+        # could otherwise inject ``/`` / ``?`` / ``#`` into the Graph URL
+        # path and cause a misrouted request. Reject anything that doesn't
+        # match the GUID shape before formatting it into the chat ID.
+        aad_object_id = from_user.get("aadObjectId")
+        if (
+            aad_object_id
+            and self._app_id
+            and not base_channel_id.startswith("19:")
+            and state
+            and _AAD_OBJECT_ID_PATTERN.fullmatch(aad_object_id)
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/adapters/teams/adapter.py:348]: Fixed in f99de08.

`re.fullmatch` raises `TypeError` on non-string input, which would have escaped `_cache_user_context` → `handle_webhook` and turned a malformed `from.aadObjectId` (int/dict/list/etc.) into a 500. Added an `isinstance(aad_object_id, str)` gate before the regex so non-conforming activities are silently skipped (no DM caching), matching the existing defense-in-depth contract.

Regression test added in `tests/test_teams_coverage.py::test_cache_user_context_handles_non_string_aad_object_id` covering `int`, `dict`, `list`, `None`, and `bool` shapes. Verified the test raises `TypeError: expected string or bytes-like object, got 'int'` against the pre-fix code and passes after.

Validation: pyrefly 0 errors, ruff clean, full suite 3691 passed (only the pre-existing unrelated `test_github_webhook::test_throws_when_no_auth` failure remains).

---
_Generated by [Claude Code](https://claude.ai/code)_
